src/CanvasContainer.tsx

```diff
@@ -2,171 +2,400 @@ import React, { CSSProperties, useEffect, useRef } from 'react';
+}
+
+const dragSessionModel = createModel(
   {
-    startPoint: { x: 0, y: 0 },
-    dragPoint: { x: 0, y: 0 },
-    dx: 0,
-    dy: 0,
-    embed: undefined as EmbedContext | undefined,
+    session: null as DragSession | null,
+    ref: null as React.MutableRefObject<HTMLElement> | null,
   },
   {
     events: {
-      LOCK: () => ({}),
-      RELEASE: () => ({}),
-      GRAB: (point: Point) => ({ point }),
-      DRAG: (point: Point) => ({ point }),
-      UNGRAB: () => ({}),
+      DRAG_SESSION_STARTED: ({ pointerId, point }: DragSession) => ({
+        pointerId,
+        point,
+      }),
+      DRAG_SESSION_STOPPED: () => ({}),
+      DRAG_POINT_MOVED: ({ point }: Pick<DragSession, 'point'>) => ({ point }),
     },
   },
 );
 
-const dragMachine = dragModel.createMachine({
-  initial: 'checking_embedded_mode',
-  states: {
-    checking_embedded_mode: {
-      always: [
-        {
-          target: 'in_embedded_mode',
-          cond: (ctx) => !!ctx.embed?.isEmbedded && !ctx.embed.pan,
+const dragSessionTracker = dragSessionModel.createMachine(
+  {
+    preserveActionOrder: true,
+    initial: 'idle',
+    states: {
+      idle: {
+        invoke: {
+          id: 'dragSessionStartedListener',
+          src:
+            ({ ref }) =>
+            (sendBack) => {
+              const node = ref!.current!;
+              const listener = (ev: PointerEvent) => {
+                const isLeftButton = ev.button === 0;
+                if (isLeftButton) {
+                  sendBack(
+                    dragSessionModel.events.DRAG_SESSION_STARTED({
+                      pointerId: ev.pointerId,
+                      point: {
+                        x: ev.pageX,
+                        y: ev.pageY,
+                      },
+                    }),
+                  );
+                }
+              };
+              node.addEventListener('pointerdown', listener);
+              return () => node.removeEventListener('pointerdown', listener);
+            },
         },
-        'released',
-      ],
-    },
-    in_embedded_mode: {},
-    released: {
-      meta: {
-        cursor: 'default',
-      },
-      invoke: {
-        src: 'invokeDetectLock',
-      },
-      on: {
-        LOCK: 'locked',
-      },
-    },
-    locked: {
-      initial: 'idle',
-      entry: ['disableTextSelection'],
-      exit: ['enableTextSelection'],
-      on: { RELEASE: 'released' },
-      invoke: {
-        src: 'invokeDetectRelease',
-      },
-      states: {
-        idle: {
-          meta: {
-            cursor: 'grab',
+        on: {
+          DRAG_SESSION_STARTED: {
+            target: 'active',
+            actions: actions.forwardTo(SpecialTargets.Parent),
           },
-          on: {
-            GRAB: {
-              target: 'grabbed',
-              actions: dragModel.assign({
-                startPoint: (_, e) => e.point,
-                dragPoint: (_, e) => e.point,
-              }),
+        },
+      },
+      active: {
```

Comment:
We should add the functionality where it disables text selection when active, see the old changes in this file. Adding them on the entry/exit here would make sense?

Reply:
from the description of the PR:
> removed disabling/enabling of the text selection. I've tried hard (experimented with different interactions with that and without that) but I couldn't see what it has been fixing. It might be that it was needed at some stage but is no longer needed. I could also just be wrong and maybe I've missed what exactly it has been fixing. If you find anything broken because of that it will be very easy to bring it back again - I would appreciate you testing this out

Reply:
Gotcha - yes, I did find that text selection was flickering when I was dragging in certain spots.

Reply:
Want a loom? You can just copy the previous implementation to add it back in.

Reply:
Ye - I would love a loom of this to understand this interaction better. I will re-add this before merging this PR.

Reply:
The loom from Matt: https://www.loom.com/share/9a14e0fcdb1f4324b973ba2727af897d

I still can't quite repro it myself but this means that those lines related to disabling & enabling text selection have to come back. Gonna do that in a sec